zbus/src/address/transport/mod.rs

```diff
@@ -52,6 +54,8 @@ use std::os::linux::net::SocketAddrExt;
     feature = "tokio-vsock"
 ))]
 pub use vsock_transport::Vsock;
+#[cfg(all(unix, feature = "tokio"))]
```

Comment:
* I don't see any reason for this to be specific to tokio feature. We even have a runtime abstraction for launching processes (it's currently only enabled for macos but it should be easy to change to for all unix systems). Talking of which, this is very similar to `Launchd` address. Please have a look into that for inspiration.
* Unlike launchd address, this address type is not connectable but only listenable. So it should only be enabled when `server` feature is enabled.

Reply:
Now the only real thing remaining is making the code make use of async abstractions to make it work on all runtimes.

Reply:
I would love to make this work for both async runtimes and in fact my first attempt was to implement this for the `async-io` runtime.

However, I wasn't able to find an *easy* way to spawn an async task from the `UnixExec::connect` method _without_ `tokio`.

I did find the [`Executor::spawn`](https://github.com/dbus2/zbus/blob/97813b0b3ec6ad6b774b1ee927c926fdd4aef821/zbus/src/abstractions/executor.rs#L38) method which does this, but that requires access to an `executor` reference which is not available in `Transport::connect` *in the current implementation*. That doesn't mean the code couldn't be refactored to pass an executor, it's just that it wasn't as straightforward as just using `tokio::io::spawn`.

And that was the main reason why this transport is only implemented for the `tokio` feature initially.

Reply:
> . That doesn't mean the code couldn't be refactored to pass an executor, it's just that it wasn't as straightforward as just using `tokio::io::spawn`.

Right and we should. We wouldn't become tokio-specific just for this feature, that's for sure. :) 

Thinking about this, I think actually it would make sense for the actual transport to be a [`Socket`](https://docs.rs/zbus/latest/zbus/connection/socket/trait.Socket.html) implementation. Then instead of having to spawn a task to pipe the data, you'll direct read and write from the stdio of the command.

Reply:
oh and btw, `main` branch just entered API break phase. If you need some API to change, this would be the perfect time to ask for that. Although if it's not a trivial change, I'm afraid I won't have the time to implement it myself. :)

Reply:
The latest revision supports both async runtimes. The new implementation required a) passing the `executor` reference as an argument to a few functions and b) adding `async-net` as a dependency.

> Thinking about this, I think actually it would make sense for the actual transport to be a [Socket](https://docs.rs/zbus/latest/zbus/connection/socket/trait.Socket.html) implementation. Then instead of having to spawn a task to pipe the data, you'll directly read and write from the stdio of the command.

:+1: This is _exactly_ the kind of feedback I was looking for when I decided to open this pull-request early, so thanks for that.

You're absolutely right, I think `Socket` might be the better abstraction for this. That being said, I get the feeling that the kind of logic that will be required for implementing this piping *properly* is fraught with peril and I don't fully trust myself to get it right.

I might give it a shot though, just to see what it would look like, but can't promise that something will come out of it.

Reply:
> The latest revision supports both async runtimes.

Awesome!

> The new implementation required a) passing the `executor` reference as an argument to a few functions

This is one of the reasons I suggested a `Socket` impl. If you don't need to spawn a task, you don't need any executor then.

> and b) adding `async-net` as a dependency.

Given that other transport types (who also need to connect) don't need `async-net`, I'm pretty sure we don't need it here either. You need to wrap the `std::os::unix::net::UnixStream` in `async_io::Async` like we do in all other places (unix streams are the most important transport for us even).

Also, you don't need any stream if you go with the `Socket` approach.

> You're absolutely right, I think `Socket` might be the better abstraction for this. That being said, I get the feeling that the kind of logic that will be required for implementing this piping _properly_ is fraught with peril and I don't fully trust myself to get it right.

That's the thing, you won't need a pipe at all AFAICT. IMO it would actually be simpler than what you're doing currently. If you look at the existing impls of `Socket`, you'll find that they're not super involved.

Reply:
> I might give it a shot though, just to see what it would look like, but can't promise that something will come out of it.

Looking at your work so far, I'm 100% confident you can do it easily. However, you're new to the project so I can understand if the low-level `Socket` intimidates you. If you need any help in implementing `Socket` (or anything), please feel free to reach out and ask (we have a matrix channel, which would be best for any questions you might have).

Knowing that `Socket` is most likely the best way forward and that going towards it would make at least 2 changes in this PR, redundant, I'm afraid I'll have to insist on going (or at least trying to) that route in this PR.

Reply:
Makes sense. Thanks for the feedback and I fully agree that the `Socket` implementation seems the better approach. I'll give that a try and ask for help if I get stuck.

Reply:
A new revision has been pushed that is using a `Socket` implementation which no longer requires passing an `executor` along various functions. I think it's looking much better now, thanks again for the feedback and guiding the solution in the right direction.

PS: The `async-net` dependency has been dropped too.

Reply:
> A new revision has been pushed that is using a `Socket` implementation which no longer requires passing an `executor` along various functions.

Nice!

> I think it's looking much better now, thanks again for the feedback and guiding the solution in the right direction.

You're most welcome. Thank you also for going the extra mile to get things done in the best way.

> PS: The `async-net` dependency has been dropped too.

cool. :)